### DISCUSSION TITLE
Could postgresql's Full text search potentially eliminate some of the custom text search implementation?

### DISCUSSION THREAD

**Discussion by nspaeth** [osm-search/Nominatim#2095] [Q&A] [answered]
What functionality does Postgres's FTS implementation need to be an adequate replacement for the text search components of Nominatim?

Would offloading this to the database improve performance, storage usage, or maintenance?

**Comment by lonvia** [comment, marked as answer]
I don't think FTS can ever fully replace the current search mechanism because there are some fundamental differences between full-text search and search for addresses. At a first glance an address may not look very different to a sentence: there are fixed elements which are expected to come in a well-defined order (from house number to country or the other way around). But that is as far as the similarity goes.

Some points that are difficult:

* Addresses have a very different grammar. That means that the usual handling of stop words, stemming etc. does not work very well.
* Addresses contain lots of non-words: housenumbers, street references, postcodes etc. They play not well with FTS.
* Geographic terms tend to not follow the usual spelling rules of the language. Some are names that have been around unchanged for hundreds of years. Others are bad translations from neighbouring languages.
* Geographic terms tend to be abbreviated a lot (street -> st, First -> 1st), a lot more than you find in a continuous text.
* We have a lot of translations for some places. Search must not only handle the multi-lingual case but also the mixed-lingual case. It is not uncommon for people to use the local name for the town part of the address but the translated version of the country part.

The search component of Nominatim is mostly busy working through all the combinations of matches that are possible due to these peculiarities and find the most likely match. The actual text search in the database is relatively quick in the end.

There is one point where Nominatim circumvents text search: every term of your query is converted into a internal word ID and the text search is done against the IDs rather than against a string. I haven't really ever thought about it but it might be possible to get rid of these intermediate IDs and search directly via FTS indexes )after doing all the normalisation we are doing already to interpret the address). It is an interesting thought. I can't say what the impact of that would be.

Regarding storage and maintenance: I'd like to emphasize that a large part of Nominatim's database is needed for computing the address terms of each place, aka preprocessing. So most of the optimisations in the last years have gone into that part because it was the lower hanging fruit.

**Reply by nspaeth** [reply]
> There is one point where Nominatim circumvents text search: every term of your query is converted into a internal word ID and the text search is done against the IDs rather than against a string. I haven't really ever thought about it but it might be possible to get rid of these intermediate IDs and search directly via FTS indexes )after doing all the normalisation we are doing already to interpret the address). It is an interesting thought. I can't say what the impact of that would be.

This is actually the part I was meaning. The different interpretations and address parts would still need to be managed by `SearchDescription`, but perhaps the scaffolding around searching for words (e.g. "burger king") could be done with FTS? Would this allow the removal of the `word` table, or would that still be required? 

My thought was that in combination with #400, the table structure might be simplified.

Thanks, I appreciated the insight into the inner-workings!

**Reply by lonvia** [reply]
The `word` table not only saves the search terms that may appear in place names. It also contains the terms for country names, postcodes, house numbers and special phrases. We'd have to preserve them somehow. But that makes up a relatively small part of the table and it could be organised differently.

The interesting question is what happens with the `search_name` table. Currently it contains two columns with search terms. `name_vector` contains all names of the place plus all the words that make up the name. Using FTS we could at least get rid of the words and just always do a full name matching on the remaining terms. `nameaddress_vector` also contains names and words but of all address parts. This is a bit trickier to match against the query terms. When a query doesn't contain commas, then we don't have any hint where the bounds of each address part is. That makes it difficult to match against full terms. Nominatim works around this by simply searching only by the words of the query and then reranking the result by the number of exact matches. We could of course do the same when using FTS but then we wouldn't gain much. Maybe there is some other clever way to match our query against the address. In the end, what I would like to see come out of it, is a way to reduce the size of the indexes on `search_name`. They have grown to a size where I'm worried that we start to loose a lot of performance soon, if the OSM keeps growing at the current speed.

It definitely is an interesting line of thought and something worth experimenting with.
